src/diagnostics.jl

```diff
@@ -55,25 +52,174 @@ function velocity_div!(grid::RegularCartesianGrid, u, v, w, div)
+@hascuda function gpu_accumulate_xy!(Rxy::CuDeviceArray{T}, Rx, data1, data2, op::Function) where T
+    lvl, lvls = blockIdx().y,  gridDim().y
+    col, cols = blockIdx().x,  gridDim().x
+    row, rows = threadIdx().x, blockDim().x
+
+    if lvl <= lvls && col <= cols && row <= rows
+        shmem = @cuDynamicSharedMem(T, 2*rows)
+        shmem[row] = _prod(data1, data2, row, col, lvl)
+        sync_threads()
+
+        # parallel reduction
+        pin, pout, offset = 1, 0, 1
+        while offset < rows
+            pout = 1 - pout
+            pin = 1 - pin
+
+            if row > offset
+                shmem[pout * rows + row] = op(shmem[pin * rows + row], shmem[pin * rows + row - offset])
+            else
+                 shmem[pout * rows + row] = shmem[pin * rows + row]
+            end
+
+            sync_threads()
+            offset *= UInt32(2)
+        end
+
+        shmem[pin * rows + row] = shmem[pout * rows + row]
+        sync_threads()
 
-    velocity_div!(model.grid, u, v, w, div)
-    min_div, mean_div, max_div = minimum(div), mean(div), maximum(div)
+        # write results
+        if row == rows
+            Rx[1, col, lvl] = shmem[row]
+        end
+        sync_threads()
 
-    if max(abs(min_div), abs(max_div)) >= diag.warn_threshold
-        t, i = model.clock.time, model.clock.iteration
-        println("time = $t, iteration = $i")
-        println("WARNING: Velocity divergence is high! min=$min_div, mean=$mean_div, max=$max_div")
+        if col == cols
+            sum = 0
+            for j in 1:cols
+                sum += Rx[1, j, lvl]
+            end
+            Rxy[1, 1, lvl] = real(sum)
+        end
     end
 
-    if max(abs(min_div), abs(max_div)) >= diag.abort_threshold
-        t, i = model.clock.time, model.clock.iteration
-        println("time = $t, iteration = $i")
-        println("Velocity divergence is too high! min=$min_div, mean=$mean_div, max=$max_div. Aborting simulation.")
-        exit(1)
+    return
+end
+
+####
+#### Horizontally averaged vertical profiles
+####
+
+mutable struct VerticalProfile{P, F, I, T} <: Diagnostic
+      profile :: P
+        field :: F
+    frequency :: I
+     interval :: T
+     previous :: Float64
+end
+
+function VerticalProfile(model, field; frequency=nothing, interval=nothing)
+    freq, int = validate_interval(frequency, interval)
+    profile = zeros(model.arch, model.grid, 1, 1, model.grid.Nz)
+    VerticalProfile(profile, field, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, P::VerticalProfile{<:Array})
+    P.profile .= mean(data(P.field), dims=[1, 2])
+end
+
+@hascuda function run_diagnostic(model::Model, P::VerticalProfile{<:CuArray})
+    Nx, Ny, Nz = P.field.grid.Nx, P.field.grid.Ny, P.field.grid.Nz
+    sz = 2Nx * sizeof(eltype(P.profile))
+    @cuda threads=Nx blocks=(Ny, Nz) shmem=sz gpu_accumulate_xy!(P.profile, model.poisson_solver.storage, P.field.data, nothing, +)

```

Comment:
Hmm. Should we move the `storage` array to `model`, and change the `poisson_solver` so that it uses it? Is this function fragile because `poisson_solver` might not always have a field `storage`?

Reply:
Ok --- so we compute a horizontal sum / average by passing one field as `nothing`, and the operation as `+`? I'm worried this is a little obscure --- in principle, I'd think you just need to pass the output array and the field that you want to sum / average over to get this result.